Fixes for defects detected by Coverity

peter-b opened the pull request.

peter-b committed:
[CID 20320] libfoundation: MCProperListRemoveElements(): Unchecked return value

peter-b committed:
[CID 16993] libfoundation: MCFiltersDecompress(): useless call

The MCSwapInt16HostToBig() (etc.) functions don't modify their
argument; they return the swabbed value.  Also, MCFiltersDecompress()
was using them the wrong way round (i.e. needed big->host not
host->big).

Coverity-ID: 16993
Coverity-ID: 16607

peter-b committed:
[CID 16596] libfoundation: MCFiltersCompress(): Useless call.

The MCSwapInt32HostToBig() function doesn't modify its argument, but
returns the swabbed value.

peter-b committed:
[CID 16362] libfoundation: MCFiltersUrlEncode(): Resource leak.

Unsuccessful attempts to avoid leaking in this function were made in
commits e3968472289b and 6acad439ec63.  This time, just use an auto
array.

Coverity-ID: 16362
Coverity-ID: 16367

peter-b committed:
[CID 14960] libfoundation: MCProperListAppendList(): Unchecked return value

peter-b committed:
[CID 15101] libfoundation: MCProperListInsertList(): Unchecked return value

peter-b committed:
[CID 15083] libfoundation: MCStringGetNativeCharPtr(): Unchecked return value

peter-b committed:
[CID 15152] libfoundation: MCStringGetCharPtr(): Unchecked return value

peter-b committed:
[CID 16122] libfoundation: MCStringUnmapTrueWordIndices(): Resource leak

peter-b committed:
[CID 16521] libfoundation: MCStringMapTrueWordIndices(): Resource leak

peter-b committed:
[CID 16750] libfoundation: MCStringUnmapIndices(): Resource leak

peter-b committed:
[CID 16786] libfoundation: MCStringMapIndices(): Resource leak

peter-b committed:
[CID 14849] libfoundation: MCTextFilter_NormalizeNFC: Incorrect deallocator

Coverity-ID: 14849
Coverity-ID: 14852

peter-b committed:
[CID 16600] libfoundation: __MCTypeInfoHash(): Missing pointer conversion

The MCStringRef *field* in the type info structure should be hashed,
not the contents of the MCString itself.

peter-b committed:
[CID 15880] libfoundation: MCUnicodeGetProperty(): Missing break in switch

peter-b committed:
[CID 16728] libfoundation: MCUnicodeGetProperty(): Incorrect overflow checks

When a signed integer is compared with an unsigned integer, it is
converted to unsigned *before the comparison*.  Therefore, this
expression is never true:

    -1 < 0U

peter-b committed:
[CID 16119] configureSerialPort(): Dereference after null check.

peter-b committed:
[CID 16165] MCLinuxDesktop::OpenDevice(): Dereference null return value

Coverity-ID: 15206
Coverity-ID: 16165
Coverity-ID: 16758

peter-b committed:
[CID 14868] libscript: MCBitwisEvalBitwiseShift(): Ensure positive shift.

Shifting by a negative amount invokes undefined behaviour.

livecodesebastien commented on libfoundation/src/foundation-unicode.cpp:
```diff
@@ -378,7 +378,7 @@ bool MCUnicodeGetProperty(const unichar_t *p_chars, uindex_t p_char_count, MCUni
                 t_prop = MCUnicodeGetIntegerProperty(t_char, p_property);
-                if (t_prop < UINT8_MIN || t_prop > UINT8_MAX)
+                if (t_prop < 0 || t_prop > UINT8_MAX)
                     return false;
```
That fixes the issue with `-1 < 0u` returning false - but more broadly, all the UINT*_MIN could be changed from `0u` to `0`, to avoid the operands cast like here.

livecodesebastien replied:
Keeping unsigned range limits unsigned is probably better in the end

runrevmark commented on libfoundation/src/foundation-string.cpp:
```diff
@@ -1065,7 +1065,8 @@ uindex_t MCStringGetLength(MCStringRef self)
 {
     if (__MCStringIsIndirect(self))
-        __MCStringResolveIndirect(self);
+		if (!__MCStringResolveIndirect(self))
```
The GetCharPtr() and GetNativeCharPtr() method present a slight problem as posed. They are not meant to throw, instead returning nil if they can't satisfy the request.

However, __MCStringResolveIndirect(), and I'm guessing __MCStringUnnativize() might both throw out of memory.

The GetNativeCharPtr() call is pretty easy to fix - if indirect, and if indirect string is native then return it's native char buffer (we will need to check that the return pointer from GetNativeCharPtr() is not used across any calls to MCString functions).

The GetCharPtr() one is slightly more tricky - I think we currently assume that it always returns a non-nil value... But due to the potential for unnativization (depending on how that's implemented), it might fail.

runrevmark commented on libfoundation/src/foundation-filters.cpp:
```diff
@@ -340,19 +340,22 @@ static const char *url_table[256] =
     // SN-2014-11-13: [[ Bug 14015 ]] We don't want to nativise the string,
     // but rather to encode it in UTF-8 and write the bytes (a '%' will be added).
-    if (!MCStringConvertToUTF8(p_source, t_utf8_string, l))
+	if (!MCStringConvertToUTF8(p_source,
```
I'm pretty sure we have an MCAutoStringRefAsUTF8String class which makes this is a little neater.

runrevmark commented on libfoundation/src/foundation-string.cpp:
```diff
@@ -1432,6 +1434,7 @@ bool MCStringMapIndices(MCStringRef self, MCBreakIteratorType p_type, MCLocaleRe
     if (t_start == kMCLocaleBreakIteratorDone)
     {
         r_out_range = MCRangeMake(MCStringGetLength(self), 0);
+		MCLocaleBreakIteratorRelease (t_iter);
```
There is a general MCAuto class for things which have a 'FREE' method (which is what MCLocaleBreakIteratorRelease) is - MCAutoCustomPointer.

peter-b committed:
Merge branch 'develop' into coverity

peter-b committed:
[CID 16362] libfoundation: MCFiltersUrlEncode(): Use MCAutoStringRefAsUTF8String

Simplifies the fix proposed in commit 49547cf7485d.

Coverity-ID: 16362
Coverity-ID: 16367

peter-b committed:
[CID 16750] libfoundation: MCStringUnmapIndices(): Use MCAutoCustomPointer.

Use MCAutoCustomPointer to manage the iterator pointer, avoiding the
need for explicit cleanup.

peter-b committed:
[CID 16786] libfoundation: MCStringMapIndices(): Use MCAutoCustomPointer.
    
Use MCAutoCustomPointer to manage the iterator pointer, avoiding the
need for explicit cleanup.

peter-b commented:
Hi @runrevmark, thanks for the feedback. I've made most of the suggested changes, but would it make sense to merge these changes now pending a plan for a more comprehensive rework of `GetCharPtr()` and `GetNativeCharPtr()`?

runrevmark commented:
@peter-b: Yes, as discussed in the office - the changes to GetCharPtr() and GetNativeCharPtr() you've done here are more 'correct' than previously; but further work is needed to correct the flaws they have (when running out of memory). This patch is fine for now.

@Reviewed-by: @runrevmark.
